MaxText/maxtext_utils.py

```diff
@@ -124,3 +126,53 @@ def calculate_tflops_prefill(num_model_parameters, prefill_length, config, log=T
           f'\t\tCausal attention TFLOPs: {causal_attention_tflops} ',
           f'({100 * causal_attention_tflops/total_tflops:.2f})% of Total')
   return total_tflops, learnable_weight_tflops, causal_attention_tflops
+
+
+def calc_not_sharded_params(shard,
+                      expected_per_device_num_param,
+                      key,
+                      pspec,
+                      num_devices_sharded
+                      ):
+  """Returns the number of params that aren't sharded."""
+  new_num_p = np.prod(shard.data.shape)
+
+  if new_num_p == expected_per_device_num_param:
+    print(f'Input is sharded over {num_devices_sharded} devices as expected!')
+  else:
+    print(f'Expected {expected_per_device_num_param} params but got {new_num_p}')
+    print(f'Off by a factor of {new_num_p // expected_per_device_num_param}')
+    print('key path', key)
+    print('pspec', pspec)
+    if jax.process_index() == 0:
+      return new_num_p
+  return 0
+
+def is_sharded_correctly(params):
+  """Checks whether most params are sharded accros sharding axis.
+
+  Given state params, function checks if over 99% of params are
+  sharded accross 'fsdp', 'fsdp_transpose','sequence', 'tensor' axes.
+  """
+  num_not_sharded_params = [0]
+  total_num_params = max_utils.calculate_num_params_from_pytree(params)
+
+  def check_state_shardings(key, pytree):
+    """Checks whether given pytree is sharded accross the sharding axes and
+    updates the number of not sharded params."""
+    pspec = pytree.sharding.spec
+    num_devices_sharded = 1
+    for axis in ['fsdp', 'fsdp_transpose','sequence', 'tensor']:
+      num_devices_sharded *= pytree.sharding.mesh.shape[axis]
+
+    num_params = max_utils.calculate_num_params_from_pytree(pytree)
+    expected_num_p = num_params // num_devices_sharded
+    for shard in pytree.addressable_shards:
+      num_not_sharded_params[0] += calc_not_sharded_params(
+        shard, expected_num_p, key, pspec, num_devices_sharded) // jax.local_device_count()
+
+  jax.tree_util.tree_map_with_path(
```

Comment:
I'm not really following this CR?

Can't we just:
```
total_num_params = max_utils.calculate_num_params_from_pytree(params)
total_num_params_per_chip = max_utils.calculate_num_params_per_chip_from_pytree(params) ###you can implement
product_num_mesh_axes_for_weight_sharded = ### you can implement but it is the number of devices in fsdp, fsdp_transpose, sequence and tensor -- you shouldn't have to look at any specific jax Arrays.
assert total_num_params_per_chip < 1.01*total_num_params/product_num_mesh_axes_for_weight_sharded
```

Reply:
I am not computing all the params per chip, but only not sharded params. E.g. if there is a weight of size 2048 that needs to be sharded on 4 devices and the number of params per device isn't 512, I am adding 2048 to the total number of not sharded devices. And I add 2048 only once because only process 0 does that. I can change the logic to what you have of course. That's another approach, but both work.

Reply:
Reimplemented the validation to match the logic you mentioned above. Checked the two are giving same results. Will close the comment for now. Please reopen if needed, thanks.